Move Advanced Search and saved search editing to the main window

AbeJellinek opened the pull request:
"Hackathon project" indeed! But I think this is a really nice UX improvement.

Doing a one-off search:

<img width="1319" height="805" alt="image" src="https://github.com/user-attachments/assets/43f90dd8-a0a7-48ea-92b0-6c12e1090358" />

Editing a saved search:

<img width="1319" height="775" alt="image" src="https://github.com/user-attachments/assets/f0c4b08c-4b08-42c7-87e5-108e935331c0" />

Needs one commit from https://github.com/zotero/zotero/pull/3435, which I've cherry-picked here.

Closes #5653

abaevbog committed:
keep search conditionIDs in arithmetic sequence

When conditions are removed, shift conditionIDs so that
conditionIDs always go in increments of 1 (0, 1, 2 ...).
It prevents conditionIDs from conflicting with each other
when conditions are rearranged.

Fixes: #3434

AbeJellinek committed:
Move Advanced Search and saved search editing to the main window

AbeJellinek committed:
Fix library-dependent conditions not showing correct UI

AbeJellinek committed:
Fix conditions not taking effect without typing

abaevbog commented:
> A global pseudo-library should probably be a separate PR, though. There's a lot of Zotero code that thinks it knows whether an item is editable based on the collection tree row that it's being viewed from. I guess we could cheaply sidestep that issue by preventing edits from within the global thing, but that doesn't seem like the right approach.

Yeah, absolutely. It would also require quite a few tweaks in the itemTree to display items within groups nicely. I will make a separate issue for it.

abaevbog commented:
Just a few additional things I noticed:

- when a condition is removed, the window is unexpectedly shrunk by a little bit
- when a reader tab is open, Edit > Advanced Search menu currently has no effect. We probably want to just focus the main tab then?
- because Advanced search was a standalone window for such a long time, I am afraid some people may forget that they have a collection selected in the collectionTree, try to use advanced search to find items in another collection and report on the forums that nothing is there. Did we discuss adding a collection condition at the top of the advanced search when a collection is selected? It would clarify things a bit I think.
- Should we disable advanced search in trash? It looks like trashed items are always excluded, so in trash advanced search is always empty for me

abaevbog commented:
> Sure. Though I imagine that optimization would prevent a new match from being added to the view? Not a big issue right now, since the only case in which that could happen would really be syncing. Could be a bigger issue later, once we support multiple library tabs.

It is supposed to add the row of an item that did not match the search but after `modify` does! That PR is one of the first ones I put together so it's a bit unruly and has several different features for no particular reason, including proper annotations handling. So I'll rebase, and try to consolidate some commits. Those tweaks are not essential for this UI update, so it could be merged later too.

AbeJellinek committed:
Don't resize window when condition added/removed

AbeJellinek committed:
Move back to library tab when Advanced Search shows

AbeJellinek committed:
Fix focus

AbeJellinek committed:
Show results in trash

dstillman committed:
Fix searches not matching items outside trash

dstillman committed:
Restore debounced quick search via oncommand

dstillman committed:
Reset collection condition menu if value missing after library change

dstillman committed:
Update Advanced Search tests for main-window UI

dstillman committed:
Clean up created items in search test

dstillman committed:
Add Advanced Search button to quick search field

dstillman committed:
Fix Feeds selection breaking on advanced search pane refresh

dstillman committed:
Update saved search tests in zoteroPaneTest for main-window UI

dstillman committed:
Add test for advanced search scoped to saved search

dstillman committed:
Fix results not refreshing when Advanced Search is closed

dstillman committed:
Apply advanced search filter via items view

The items view's row provider holds the CollectionTreeRow instance it
was given in changeCollectionTreeRow(), while the collection tree can
recreate its row objects when it refreshes, so getCollectionTreeRow()
can return a different object from the one the items view is using.

_refreshAdvancedSearchPane() set the advanced search on the collection
tree's current row, so once the instances diverged, opening or closing
the pane modified a row the items view never consulted, while
submitting a search -- which goes through setFilter() -- modified the
items view's own row. A previously submitted search could therefore
remain in effect, with the items list left blank, after closing the
pane or reopening it with no conditions.

Apply the filter via ItemTree::setFilter() instead, the same path used
when submitting, which sets it on the items view's row and refreshes
the list. During collection selection the newly selected row isn't in
the items view yet, so onCollectionSelected() passes it in to have the
search set on it directly.

Also fix a regression from 464b361a1e by updating the pane state
synchronously before awaiting the refresh, so that reopening the pane
while a close was still refreshing doesn't reapply the old search.

dstillman committed:
Don't filter items list until advanced search is submitted

Opening the Advanced Search pane applied its default blank condition
(Title contains "") to the items list immediately, and a blank search
isn't a no-op: it matches child items, expanding all parents in the
list to show them; it excludes items without a title; and, combined
with trash inclusion, it shows trashed duplicates of library items.

Track whether the pane's search has been submitted and only filter the
items list once it has. Opening the pane now leaves the list unchanged
until Search is clicked, as with the old Advanced Search window, and
Clear removes the filter rather than submitting the empty search. A
submitted search still carries over when switching collections, until
the pane is closed or cleared.

dstillman committed:
Fix advanced search in Feeds view

The advanced search clone was given the libraryID of the current
collection tree row's ref, but the Feeds pseudo-library doesn't have
one, so the clone kept its original library and the intersection with
the feed-items scope was always empty. Leave the library unset (null)
in that case so that the search spans all feed libraries, the same way
the feeds row's own scope search works. A library-less search can't be
clone()d, so copy it via toJSON()/fromJSON().

dstillman commented:
Merged in d2f1c56250 and 00527332c4

dstillman commented:
@yexingsha: If you have tweaks for the UI, let us know. Along with the main changes here, I moved the advanced search action into a filter button within the quick search bar. Right now that's a little conceptually messy, since it's separate from quick search, but 1) we've had the Advanced Search menu option in the quick search mode menu for a while and 2) we could make it cleaner by automatically creating advanced-search conditions from the current quick search, which would be neat (and could even work if we extended the quick search syntax to allow prefixes/tokens, etc.).

yexingsha commented:
Here is what I have so far:

- Save Search should open a dialog for name input.
	- I was wondering if we could enable Save Search in collections as well, and there the dialog will also contain a checkbox that says "Include the current collection as a condition" or something like that.
- The height of the Advanced Search area seems to be calculated weirdly. At default height, the space below the buttons is larger than what it should be (8px). When I add any new conditions or Edit Saved Search, the space below the buttons become smaller. After I exit Edit Saved Search, the Advanced Search area no longer returns to its minimal height.
- The Advanced Search area should have a 1px bottom border using `color-panedivider`.
- Various alignment and spacing issues: because some input/combo boxes (but not all) have their own margins from Firefox css, the resulting padding and spacing becomes larger than intended. It would be nice if we could patch out the margins on the individual components, but if not, we will need to fine tune the margins by row.
- Move the filter button 4px to the right (`margin-inline-start: -26px`) so it has the same margin on all sides.

AbeJellinek commented:
On macOS, the condition value field seems to extend too far down:

<img width="801" height="55" alt="image" src="https://github.com/user-attachments/assets/f7ffb5ce-b462-4e06-a973-4d60139751e9" />
